### DISCUSSION TITLE
[Kamal 2]: Container Conflicts for Staging and Production Application Running on Same Host

### DISCUSSION THREAD

**Discussion by mike-weiner** [basecamp/kamal#1110] [Q&A] [answered]
I've got a basic web application running on a DigitalOcean Droplet. I have both the production and staging instances of the application on the same host. I realize that's atypical. It's a non-critical, side project that I don't need redundancy for nor do I need to add complexity for 2 different hosts. 

I'm hitting a weird edge case that I'm looking for some advice on. 

When I've got a new build ready to go out to staging, I do a `kamal deploy -d staging` and it builds the container with the current commit (let's say that is `123`), builds the image, and exposes port `abcd` for my web application. That deploys perfectly. 

24 hours later I'll do a `kamal deploy` to push the changes out to production. Kamal re-builds the container with the same commit (`123`) but it exposes a different port, let's say `efgh` for production. That process then overwrites that image in my registry with a build that now exposes port `efgh` instead of `abcd`.

The production instance deploys fine, but if I ever need to restart the staging instance, it will go and re-pull the `123` container image which is now built to expose the incorrect port. 

Is there a better way to achieve this kind of setup to avoid this? Or is this a limitation that I'm going to need to live with?

Thank you!

**Comment by tuladhar** [comment]
Are you using same `service:` name for both your staging and production?

**Reply by mike-weiner** [reply]
Yes, I am.

**Reply by tuladhar** [reply]
The service name must be unique, you can prefix or suffix with environment name, e.g: `myapp-staging`

**Reply by mike-weiner** [reply]
Here is my config: https://github.com/mike-weiner/kenyonwx/tree/main/config

The problem that I have is that I do a `kamal deploy -d staging` and that pushes both a `latest-staging` and a commit (`57c85d4062d5de024719fde5c4b42adf3aef3c67`) image to Docker. 

I then run `kamal deploy` and that pushes both a `latest` and a commit (`57c85d4062d5de024719fde5c4b42adf3aef3c67`) image to Docker that overwrites the previous `57c85d4062d5de024719fde5c4b42adf3aef3c67` tagged image. And that changes the exposed port in the image. 

Kamal uses the commit image to deploy. So this overwriting of the commit image tag is a problem as the port changes between my staging and production environment.

**Reply by tuladhar** [reply]
Hmm...in that case, this might help.
```
kamal deploy -d staging --version staging
```

**Reply by mike-weiner** [reply]
Thanks, @tuladhar. That command did work. 

To be clear, I think my use case is not very common. Most people wouldn't build a different container image that exposes a different port for their staging and production instances. And most people would not deploy their staging and production applications onto the same box.

However, since I deploy both to the same box, I need different container ports exposed for each build.

**Reply by djmb** [reply]
Hey @mike-weiner!

Kamal won't publish the ports for your app containers, so I think it should be ok to use the same one in production and staging and to deploy them to the same box. Does that work?

**Reply by mike-weiner** [reply]
Hi, @djmb 👋🏻. 

I'm not sure that I'm following. Or, my original description wasn't super clear. 

My understanding is that Kamal Proxy exposes ports 80 and 443 to be able to forward traffic to the correct backend applications. My container image exposes a port for my web application to listen on. However, you can't have multiple Docker containers running at the same time, on the same box, that expose the same host port.

For production I'm exposing port `xxxx` in my container image. In my Kamal configuration I am doing the following:

```
proxy:
  ...
  app_port: xxxx
```

For staging, I'm exposing port `yyyy` in my container image. Then in my Kamal configuration for my staging destination I have:

```
proxy:
  ...
  app_port: yyyy
```

When I had a commit that I pushed to staging (followed by production), I would do the following: 

1. `kamal deploy -d staging`

This would push 2 images out to my registry. One with a tag of `latest-staging` and one with the tag of the Git commit, i.e. `abcd`. Those images were built using my `Dockerfile.staging` that exposed port `yyyy`. 

The container would correctly start on my box and would be exposing port `yyyy` and it uses the `abcd` tagged image. 

2. I would then run `kamal deploy` when I was ready to push my changes to production. 

This would push 2 images out to my registry. One with a tag of `latest` and one with the **same** tag of the Git commit, i.e. `abcd`. The Docker image with the tag of the Git commit that once exposed `yyyy` is overwritten and now exposes port `xxxx`. 

The container would correctly start on my box and would be exposing port `xxxx`. 

Now, if I went to restart my staging deployment, it would re-pull that image tagged with the Git commit of `abcd`, which is now incorrectly exposing port `yyyy` instead of `xxxx` since it was re-built to deploy to production.

---

Other than the solution provided by @tuladhar above, the only other solution that I can think is if I could tweak the `docker run` command that Kamal is doing. That way I could keep all of my container images exposing a single port of `zzzz` but in the `docker run` commands I could map `xxxx` on the host to `zzzz` in the container and `yyyy` on the host to `zzzz` in the container.

**Reply by djmb** [reply]
Exposing and publishing are different things in Docker.

Kamal-proxy publishes to ports 80 and 443 which binds them on the host, so one else can use them on that host.

Exposing ports makes them available to other containers on the same network, so they can be accessed that via `<container_id>:<app_port>`. There should be no need to use different ports as they are not bound on the host. Unless there's something else in your configuration that's causing an issue?

**Reply by mike-weiner** [reply]
@djmb - Hmmm... there shouldn't be. In the next couple days I will create a basic example of my problem and follow up here. I'm hoping having something a bit more concrete to point at might help here. 

Thanks!

**Reply by mike-weiner** [reply]
@djmb - Thanks for bearing with me. I've got a basic re-create of the behavior I'm seeing. 

I'm using [this repo of mine](https://github.com/mike-weiner/upbeat/tree/main) for this test. It is a basic Next.js web application that does nothing other than provide `/up`.

I spun up a brand new DO Droplet for this test. So, there's nothing else running on this VM and it is heavily over-provisioned on compute power.

I then ran `kamal setup` for that repo above, and I was successfully able to visit `up.springcreek.io/up` and saw my application running as a basic sanity check. 

I then ran `kamal remove app` to tear out that application.

I then ran `kamal deploy -d staging` to get my staging instance running. This would mimic a "real world" scenario as any new changes would first be going out to stage. That `kamal deploy -d staging` command completed successfully and I was able to visit `up2.springcreek.io/up` successfully.

I then ran `kamal deploy` to get my production instance re-started. That `kamal deploy` command completed successfully and I was able to visit `up.springcreek.io/up`. 

However, I then went and refreshed `up2.springcreek.io/up`, and that was returning a Bad Gateway 502 (🙃). 

I jumped onto the Droplet itself and I only see 2 containers running (my production web app and the `kamal-proxy` container). There should have been 3 containers running (my staging web app, the production web app, and the `kamal-proxy` container).

Per `docker`, my staging container was killed right as my production container was started:

```
~$ docker ps -a
CONTAINER ID   IMAGE                                                                                   COMMAND                  CREATED          STATUS                      PORTS                                                                      NAMES
588e14e8dddf   registry.hub.docker.com/michaelweiner/upbeat:7d23aa9dd7985a3e592fa455f2fe8a1f4a4dfcf2   "docker-entrypoint.s…"   24 seconds ago   Up 23 seconds               3000/tcp                                                                   upbeat-web-7d23aa9dd7985a3e592fa455f2fe8a1f4a4dfcf2
c8ddf8925b38   3b5705b1fc99                                                                            "docker-entrypoint.s…"   2 minutes ago    Exited (1) 20 seconds ago                                                                              upbeat-web-staging-7d23aa9dd7985a3e592fa455f2fe8a1f4a4dfcf2
d32641eac5c4   basecamp/kamal-proxy:v0.7.0                                                             "kamal-proxy run"        54 minutes ago   Up 54 minutes               0.0.0.0:80->80/tcp, :::80->8
```

The logs from my staging instance show:

```
~$ docker logs c8ddf8925b38

> upbeat@1.0.0 start
> next start

  ▲ Next.js 14.2.13
  - Local:        http://localhost:3000

 ✓ Starting...
 ✓ Ready in 381ms
npm error path /app
npm error command failed
npm error signal SIGTERM
npm error command sh -c next start
npm error A complete log of this run can be found in: /root/.npm/_logs/2024-10-26T02_35_43_956Z-debug-0.log
```

So, something sent a `SIGTERM` into the container running my staging deployment. It sure seems to have been `kamal` that did that. 

I've left my machine in this state. So, `https://up2.springcreek.io/up` remains broken right now. (**Note:** If I were to run `kamal deploy -d staging` that would get my staging deployment running again and it would not touch my production deployment.)

This is what originally (and incorrectly) sent me down the path of assuming that this was caused by having two containers exposing the same port (in this case that is `3000`). What's interesting is that I do **not** see this behavior when my staging deployment exposes a different port than my production deployment. 

What's concerning to me is that none of my `kamal deploy ...` commands failed, but my staging instance was completely tore down.

Is this a configuration issue on my end? 

If we want to take this to a separate conversation, let me know and I'm happy to break this out somewhere else.

Thanks!

**Reply by mike-weiner** [reply]
@djmb ^ Friendly bump on the above.

**Reply by djmb** [reply]
Hey @mike-weiner!

Kamal runs "docker stop" as part of the deploy, so if it is killing the staging container then that's when it would be happening. It shouldn't be doing this though, but obviously something is killing the container.

Could you repeat these steps and look for "docker stop" in the logs? The command will look something like:

```
docker container ls --all --filter name=^app-web-123$ --quiet | xargs docker stop
```

If whatever you have in the `docker container ls --all --filter name=^app-web-123$ --quiet` matches your staging container then that's likely to be it.

For further debugging, there's also a `docker events` command that might track what's killing the container.

**Reply by mike-weiner** [reply]
@djmb - Found the problem.

`kamal deploy -d staging`:

```
...
Detect stale containers...
  INFO [ac667262] Running docker ps --filter label=service=upbeat --filter label=destination=staging --filter label=role=web --format "{{.Names}}" | while read line; do echo ${line#upbeat-web-staging-}; done on 159.203.151.191
  INFO [ac667262] Finished in 0.178 seconds with exit status 0 (successful).
...
```

`kamal deploy`:

```
...
Detect stale containers...
  INFO [fe423d05] Running docker ps --filter label=service=upbeat --filter label=role=web --format "{{.Names}}" | while read line; do echo ${line#upbeat-web-}; done on 159.203.151.191
  INFO [fe423d05] Finished in 0.195 seconds with exit status 0 (successful).
  INFO [2eadd5da] Running /usr/bin/env sh -c 'docker ps --latest --format '\''{{.Names}}'\'' --filter label=service=upbeat --filter label=role=web --filter status=running --filter status=restarting --filter ancestor=$(docker image ls --filter reference=registry.hub.docker.com/michaelweiner/upbeat:latest --format '\''{{.ID}}'\'') ; docker ps --latest --format '\''{{.Names}}'\'' --filter label=service=upbeat --filter label=role=web --filter status=running --filter status=restarting' | head -1 | while read line; do echo ${line#upbeat-web-}; done on 159.203.151.191
  INFO [2eadd5da] Finished in 0.256 seconds with exit status 0 (successful).
App Host: 159.203.151.191
Stopping stale container for role web with version staging-7d23aa9dd7985a3e592fa455f2fe8a1f4a4dfcf2

  INFO [05d3dee2] Running docker container ls --all --filter name=^upbeat-web-staging-7d23aa9dd7985a3e592fa455f2fe8a1f4a4dfcf2$ --quiet | xargs docker stop on 159.203.151.191
  INFO [05d3dee2] Finished in 1.281 seconds with exit status 0 (successful).
...
```

The key difference here is the `docker ps --filter label=service=upbeat --filter label=role=web --format "{{.Names}}"` when doing `kamal deploy` versus the `docker ps --filter label=service=upbeat --filter label=destination=staging --filter label=role=web --format "{{.Names}}"` when running `kamal deploy -d staging`.

I went and looked at the labels on my staging container and my production container: 

```
~$ docker inspect --format '{{ json .Config.Labels }}' 721dc3f5a52f
{"destination":"staging","role":"web","service":"upbeat"}
~$ docker inspect --format '{{ json .Config.Labels }}' 3908b053db86
{"destination":"","role":"web","service":"upbeat"}
```

```
~$ docker ps --filter label=service=upbeat --filter label=role=web --format "{{.Names}}"
upbeat-web-staging-7d23aa9dd7985a3e592fa455f2fe8a1f4a4dfcf2
upbeat-web-7d23aa9dd7985a3e592fa455f2fe8a1f4a4dfcf2
```

So, that explains why doing `kamal deploy` is killing my "stale" staging container.

That begs another question for me. Why does `kamal deploy` not add the `--filter label=destination=`? I would expect that if I don't specify a destination that `kamal` wouldn't go and change the state of containers that do have a `destination` label.

```
~$ docker ps --filter label=service=upbeat --filter label=role=web --filter label=destination= --format "{{.Names}}"
upbeat-web-7d23aa9dd7985a3e592fa455f2fe8a1f4a4dfcf2
```

**Reply by djmb** [reply, marked as answer]
Ah! That's a bug then - I'll convert this to an issue and pick it up next week. 

Thanks for digging into this 🙏

**Reply by mike-weiner** [reply]
Awesome. Thank you for helping dig into this!!

(For anyone else interested, here is the associated issue: https://github.com/basecamp/kamal/issues/1184)
